src/ata.c

```diff
@@ -301,7 +301,6 @@ static int ata_write_unlocked( int id, const void *buffer, int nblocks, int offs
 		buffer = ((char*)buffer)+ATA_BLOCKSIZE;
 		offset++;
 	}
-	if(ata_interrupt_active) process_wait(&queue);
```

Comment:
Ok, the issue appears to be that the interrupt occurs before we ever get here, so printing the interrupts:

![screen shot 2017-12-13 at 5 57 21 pm](https://user-images.githubusercontent.com/8294394/33966883-366b7d06-e02f-11e7-9bad-3db6190e0d8c.png)

So at the end, you can see we are waiting for an interrupt that has already been fired.

The solution to me seems to be checking the disk is busy before calling `process_wait`. Does that make sense?

Reply:
Not quite -- what we have here is a race condition, likely due to the fact that the virtual hardware is "too fast" -- the driver requests an action, and then the hardware interrupts before the driver even has a chance to sleep.  (Real hardware wouldn't respond that quickly.)  Checking first doesn't help the problem, because the interrupt could still occur after the check but before the sleep.

There are two ways to solve this problem:
1 - Disable all interrupts (or that specific interrupt) from the creation of the request until after process_wait.  (Note that process_switch re-enables interrupts as part of the context switch.)
2 - Accumulate a count of interrupts in the handler, so that if it comes in before process_wait is called, simply bail out without waiting.

Reply:
I should add that you still need to block interrupts briefly in #2, but only within one function while checking the accumulated count, so it's a little easier to reason about..